colcon/colcon-bundle
# Add dependency hash check during bundling

Add dependency hash check during bundling and save bundle time if dependencies have not changed.

**Testing**
pytest unittest
integration test
functional test:
1. build v2 bundle against test_workspace for the first time (~15 minutes)
1. build again and it takes 40 seconds to build since the hash matches
1. change the order of the dependencies in the package.xml, build again, and it takes the same 40 seconds since the hash matches
1. add 'roscpp' as exec dependency in the test-cmake/package.xml and build. Hash mismatch and it takes 4 minutes to build
1. build again and it takes 40 seconds since hash matches 


New commit: [e899841](https://github.com/colcon/colcon-bundle/commit/e899841e43a5fb08d19e9667759a258abecfb007)

1. Check dependencies.tar.gz existence before checking dependencies hash to avoid error
2. Add --upgrade to force check for version upgrades

Testing:

1. With existing dependencies.tar.gz and metadata files, bundling without '--upgrade' will only check package dependencies and bundling from dependencies.tar.gz
2. With existing dependencies.tar.gz and metadata files, bundling with '--upgrade' will check package dependencies and installer dependencies
3. With existing dependencies.tar.gz and metadata files, add new package in package.xml, bundling without '--upgrade' will only check package dependencies and bundling from dependencies.tar.gz
4. With existing dependencies.tar.gz and metadata files, add new package in package.xml, bundling with '--upgrade' will check package dependencies and installer dependencies, and rebuild the dependencies.tar.gz
5. Remove dependencies.tar.gz, bundling will check package dependencies and installer dependencies (which will also do the package install), and rebuild dependencies.tar.gz
